hcl/fmt.go

```diff
@@ -151,3 +164,309 @@ func (f FormatResult) Formatted() string {
+		Logger()
+
+	// We are interested on lists, ignore the rest
+	if tokens[0].Type != hclsyntax.TokenOBrack {
+		logger.Trace().Msg("not a list, returning tokens as is")
+		return tokens
+	}
+
+	if isListComprehension(tokens) {
+		logger.Trace().Msg("list comprehension, ignoring")
+		return tokens
+	}
+
+	formattedList, pos := fmtListExpr(tokens)
+	if pos != len(tokens) {
+		panic(fmt.Errorf(
+			"last pos %d != tokens len %d for tokens: %q",
+			pos,
+			len(tokens),
+			tokensStr(tokens),
+		))
+	}
+	return formattedList
+}
+
+// fmtListExpr will adjust the given list tokens so they can be formatted
+// properly. It returns the adjusted tokens and the position of the first
+// token after the list ended.
+//
+// If there is no more tokens after the end of
+// the list the returned position will be equal to len(tokens).
+func fmtListExpr(tokens hclwrite.Tokens) (hclwrite.Tokens, int) {
+	logger := log.With().
+		Str("action", "hcl.fmtListExpr()").
+		Str("tokens", tokensStr(tokens)).
+		Logger()
+
+	logger.Trace().Msg("formatting list")
+
+	elemNextPos := 0
+	newTokens := hclwrite.Tokens{tokens[elemNextPos], newlineToken()}
+	elemNextPos++
+
+	skipNls := func() {
+		_, skipped := skipNewlines(tokens[elemNextPos:])
+		elemNextPos += skipped
+	}
+
+	for {
+		skipNls()
+
+		nextTokenType := tokens[elemNextPos].Type
+
+		if nextTokenType == hclsyntax.TokenComma {
+			elemNextPos++
+			continue
+		}
+
+		if nextTokenType == hclsyntax.TokenComment {
+			newTokens = append(newTokens, tokens[elemNextPos])
+			elemNextPos++
+			continue
+		}
+
+		if nextTokenType == hclsyntax.TokenCBrack {
+			logger.Trace().Msg("reached end of list")
+			break
+		}
+
+		logger.Trace().Msg("getting next element of the list")
+
+		element, nextPos := fmtNextElement(tokens[elemNextPos:])
+		elemNextPos += nextPos
+
+		logger.Trace().
+			Str("element", tokensStr(element)).
+			Str("tokens", tokensStr(tokens)).
+			Int("elemNextPos", elemNextPos).
+			Msg("new element got")
+
+		newTokens = append(newTokens, element...)
+		// Heredocs need to be handled differently, the comma must
+		// be on the next line in this case
+		if isHeredoc(element) {
+			newTokens = append(newTokens, newlineToken())
+		}
+		newTokens = append(newTokens, commaToken(), newlineToken())
+	}
+
+	newTokens = append(newTokens, closeBracketToken())
+	elemNextPos++
+
+	// Handling ["one"][0] and things like [[0]%[0]]
+	// We can also have newlines when dealing with operations
+
+	logger.Trace().Msg("checking if formatted list has operators/index access")
+
+	skipNls()
+
+	if elemNextPos == len(tokens) {
+		logger.Trace().Msg("no more tokens, returning formatted list")
+		return newTokens, elemNextPos
+	}
+
+	// We are handling things like this:
+	// var = [[] # c1\n\n #c\n [*]]
+	// We need to keep any comments after the immediate end of the list
+	// We don't keep the extra newlines, only newlines belonging to comments themselves.
+	hasCommentOrNl := true
+	for hasCommentOrNl {
```

Comment:
Yeah I remember this one from partial eval, indeed it will be better :-)

Reply:
Ah actually I didn't use it here because I'm not skipping newlines and comments, I skip newlines + retain comments as far as the only thing that I'm finding are newlines and comments, so it won't "skip" the comments, only newlines. The generalize function only skips all tokens provided (from what I remember).

Reply:
We do have a skipTokens function now since I used it for the list comprehension with docs inside thing, but I still can't use it here I think.